### ISSUE TITLE
Conformance based on feature groups

### ISSUE THREAD

**Issue by hbagdi** [kubernetes-sigs/gateway-api#764]
**What would you like to be added**:

At the beginning, this API was focused on HTTP with plans to include other protocols.
Then L4 protocols were added in.
Overtime, we have added support for Service Mesh like use cases.
Add "conformance profile" to the API so that an implementation can support only a subset of the API. For example, an HTTP gateway, L4 gateway or a mesh gateway.

**Why this is needed**:

A single implementation may wish to focus on specific feature group and not support the entire API. This is not possible today.

**Comment by youngnick** [comment]
Some not-quite-ready-for-a-GEP thoughts from me:

Roughly, it seems like we've got three main classes of thing wanting to use this API right now:
- Ingress controllers (layer 7 load balancers/application gateways that route mainly HTTP traffic)
- Load Balancers (layer 4 load balancers that route TCP and UDP, maybe also handling some TLS)
- Mesh controllers (that provide a gateway between the inside and outside of a service mesh)

Does that sound about right to everyone?

If that is the case, we need to explain what a conformance profile would actually do. In my mind, there are a few advantages to having conformance profiles:
- Creates easy-to-understand use cases for users to know if they can use the API
- Routes should be able to move between Gateways of the same conformance profile with minimal changes
- Defining what a conformance profile means allows us to define conformance tests that implementations can use to certify that they meet the contract.
- The list is not exhaustive, we can always add more profiles later if other use cases arrive.

So, what should be in a conformance profile? I think we can actually do this in a reasonably straightforward way to begin with, and define a conformance profile as the support of certain objects in the API, something like this (YAML used because my brain is now warped into thinking in it):

```yaml
conformance:
  - ingress-controller:
      core:
      - GatewayClass
      - Gateway
      - ReferencePolicy
      - HTTPRoute
      extended:
      - TLSRoute
  - l4-lb:
      core:
      - GatewayClass
      - Gateway
      - ReferencePolicy
      - TCPRoute
      - UDPRoute
      extended:
      - TLSRoute
  - mesh:
      core:
      - GatewayClass
      - Gateway
      - ReferencePolicy
      - HTTPRoute?
```

The point here is that the profiles are additive; an implementation may support one or more, so can have an implementation that provides layer 4 and ingress-controller functions.

Obviously, there will need to be things defined for what support of a particular object means, like we've already done for HTTP, but I think this goes a long way to getting us started.

**Comment by bowei** [comment]
The names of the categories might want to change (like make it L7, L4, mesh); but the general direction +1 from me.

**Comment by robscott** [comment]
I believe this was fixed by @youngnick's work on GEP #917. Closing this one out, feel free to reopen if I missed anything.

/close
